### DISCUSSION TITLE
Complete step for 1000 wf

### DISCUSSION THREAD

**Discussion by devrebeleza** [flowsforapex/apex-flowsforapex#583] [Q&A]
Hi community.

We recently started using flowforapex to increase the functionality of our product. So far it's been great, but we're having some issues with the number of instances to close.

We created three models (one for level), the last model/level has many instances created (1000 or many more).
  
Each instance takes 2 seconds to complete, for 1000 instances it will take at least **33 minutes** to fully finish.

This is our model.

![image](https://user-images.githubusercontent.com/32907345/222217012-8935f7c3-f606-4fa5-a698-1a06b3e356df.png)

The only script that has some code in it is "_Notify Role Group Instance of Completion_". This script only send notification to the upper level that this instances it's finished.

My questions are:
1.  Is if this 2 seconds delay is correct or are we missing something? 
2. Can FlowsForApex manage a large number of instances and complete them simultaneously? (+5000) 

Thanks in advance!
Renzo

**Comment by rallen2010** [comment]
If you have three models that are all working together on the same objects, and the throughput is one transaction every 2 seconds, then I suspect that you are doing something which is trying to lock the same objects in multiple models.  A lot of the engine locking is done with a 2 second timeout - so this seems to be an interesting coincidence.  We know that Flows for APEX running on an always free tier OCI platform can process steps much faster than that, so I suspect a locking issue.  
a an you provide all three interrelated models, either in the discussion forum or privately?

**Comment by devrebeleza** [comment]
Thanks for your answer.

We discovered a blockage and we are trying to take another approach.

However I attached a zip file with the three models.

A little explanation of the process.

* wf_certif_level_1 / wf_certif_level_2 / wf_certif_level_3
Level 1 starts a certification and initiate two or more instances from level 2 (maximum instances for our product is 6).
Instances for level 2 create and start instances for level 3 (approximately 1000 for each level 2).
 
The problem starts when we want to close all level 3 instances, which depend on one level 2 instance.

To close a level 2, all instances in level 3 must be closed first. 


This is an excerpt from our code when we decide whether or not to complete a step of a level 2 instance.

`
...
 -- Check if all children instances were completed

      SELECT NVL (MAX (1), 0)
        INTO v_count
        FROM flows4apex.flow_processes
       WHERE prcs_id IN (
                SELECT user_access_prcs_id
                  FROM fsd.fsd_cer_user_roles
                 WHERE role_group_prcs_id = p_role_group_prcs_id
                   AND user_access_prcs_id <> p_user_access_prcs_id)
         AND UPPER (prcs_status) <> 'COMPLETED';

      --
      IF v_count = 0
      THEN
         --
         flows4apex.fsd_cer_pkg.move_role_group_instance_prc
                                                       (p_role_group_prcs_id,
                                                        v_cer_id,
                                                        v_company_id
                                                       );
      --
      END IF;
...
`
[WF_CERTIF.zip](https://github.com/flowsforapex/apex-flowsforapex/files/10876760/WF_CERTIF.zip)
